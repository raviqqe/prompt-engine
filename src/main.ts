export interface Example {
  input: string;
  output: string;
}

const separator = "\n\n";

export const buildPrompt = (
  description: string,
  examples: [Example, ...Example[]],
  input: string
): string =>
  [
    description,
    examples
      .flatMap(({ input, output }) => [
        buildUtterance(input, true),
        buildUtterance(output, false),
      ])
      .join(separator),
    buildUtterance(input, true),
    buildRole(false),
  ].join(separator) + "\n";

const buildUtterance = (data: string, user: boolean) =>
  `
${buildRole(user)}
${data}
`.trim();

const buildRole = (user: boolean): string =>
  `"""${(user ? "user" : "bot").toUpperCase()}`;
