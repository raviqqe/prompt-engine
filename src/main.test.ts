import { describe, expect, it } from "vitest";
import { buildPrompt } from "./main.js";

describe(buildPrompt.name, () => {
  it("builds a prompt", () => {
    expect(
      buildPrompt("foo", [{ input: "bar", output: "baz" }], "qux"),
    ).toMatchSnapshot();
  });

  it("builds a prompt with two examples", () => {
    expect(
      buildPrompt(
        "foo",
        [
          { input: "bar", output: "baz" },
          { input: "answer", output: "42" },
        ],
        "qux",
      ),
    ).toMatchSnapshot();
  });
});
