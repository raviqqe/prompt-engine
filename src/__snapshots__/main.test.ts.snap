// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildPrompt > builds a prompt 1`] = `
"foo

\\"\\"\\"USER
bar

\\"\\"\\"BOT
baz

\\"\\"\\"USER
qux

\\"\\"\\"BOT"
`;

exports[`buildPrompt > builds a prompt with two examples 1`] = `
"foo

\\"\\"\\"USER
bar

\\"\\"\\"BOT
baz

\\"\\"\\"USER
answer

\\"\\"\\"BOT
42

\\"\\"\\"USER
qux

\\"\\"\\"BOT"
`;
